File: packages/types/src/acp/content.ts
/**
 * ACP content blocks and streamed chunks (todo #350), scoped to the `text`
 * and `image` variants (spec Decision 22): both have producers
 * (`LeafContent.Text`/`Image`), while `audio`/`resource`/`resource_link`
 * stay out — no producer in any adapter pipeline (spec Decision 17).
 * Mirrors `mainframe-types/src/acp/content.rs`.
 */
import { z } from 'zod';

export type MessageId = string;

/**
 * `_meta` mirrors the schema's discipline of reserving it on essentially
 * every nested struct; Mainframe uses it for the namespaced truncation
 * marker on tool-result text (`extensions.ts`'s `TruncationMarker`).
 */
export const TextContentBlockSchema = z
  .object({
    type: z.literal('text'),
    text: z.string(),
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type TextContentBlock = z.infer<typeof TextContentBlockSchema>;

/** Schema `ImageContent`: base64 `data` + `mimeType` required, `uri` optional. */
export const ImageContentBlockSchema = z
  .object({
    type: z.literal('image'),
    data: z.string(),
    mimeType: z.string(),
    uri: z.string().optional(),
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type ImageContentBlock = z.infer<typeof ImageContentBlockSchema>;

export const ContentBlockSchema = z.discriminatedUnion('type', [TextContentBlockSchema, ImageContentBlockSchema]);
export type ContentBlock = z.infer<typeof ContentBlockSchema>;

/** One streamed item of message content — the `*_chunk` `session/update` payload. */
export const ContentChunkSchema = z
  .object({
    messageId: z.string(),
    content: ContentBlockSchema,
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type ContentChunk = z.infer<typeof ContentChunkSchema>;

File: packages/core-rs/crates/mainframe-types/src/acp/content.rs
//! ACP content blocks (schema `ContentBlock`) and streamed chunks
//! (`ContentChunk`), scoped to the `text` variant — an explicit deviation
//! (spec Decision 17): `image` is blocked on the delta engine's single-text-
//! block chunk grammar (deferred with the partial-message phase), and
//! `audio`/`resource`/`resource_link` have no producer in any adapter
//! pipeline. Adding a variant is additive to this enum.

use serde::{Deserialize, Serialize};
use serde_json::Value;

pub type MessageId = String;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ContentBlock {
    Text { text: String },
}

/// One streamed item of message content (schema `ContentChunk`) — the
/// `user_message_chunk` / `agent_message_chunk` / `agent_thought_chunk`
/// `session/update` payload.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContentChunk {
    pub message_id: MessageId,
    pub content: ContentBlock,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "_meta")]
    pub meta: Option<Value>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn text_block_round_trips() {
        let v = json!({ "type": "text", "text": "hello" });
        let parsed: ContentBlock = serde_json::from_value(v.clone()).unwrap();
        assert_eq!(serde_json::to_value(&parsed).unwrap(), v);
    }
}
